### PR TITLE
feat(migration): migrate lb4-llm-chat-extension from LangGraph to Mastra 

### PR DESCRIPTION
**Pull request by piyushsinghgaur1**
This pull request introduces a significant refactor to support the Mastra workflow engine, improve compatibility with Mastra's language model interface, and enhance the testability and maintainability of the codebase. The changes include dependency updates, new service bindings, expanded test coverage, and new utility methods for chat message handling. The most important changes are grouped below:

**Mastra Integration and Dependency Updates**
- Added `@mastra/core` as a dependency in `package.json` to enable Mastra workflow and LLM support.
- Registered `WorkflowRunner` as a provider in the component and replaced usage of `ChatGraph` with `WorkflowRunner` in integration tests and component registration. [[1]](diffhunk://#diff-8725d3d09c584970a100d25ac569f0f711a21b2dc0e72cb50c725aabe5fb3c0eR60) [[2]](diffhunk://#diff-8725d3d09c584970a100d25ac569f0f711a21b2dc0e72cb50c725aabe5fb3c0eR92-R93) [[3]](diffhunk://#diff-1508ed91a3800267548c8dbd513552b1472bc2d4a2740946d13c27403261e5dcL1-R40) [[4]](diffhunk://#diff-1508ed91a3800267548c8dbd513552b1472bc2d4a2740946d13c27403261e5dcL33-R88) [[5]](diffhunk://#diff-1508ed91a3800267548c8dbd513552b1472bc2d4a2740946d13c27403261e5dcL153-R121)

**Service Bindings for Mastra LLMs**
- Introduced new binding keys `MastraChatLLM` and `MastraFileLLM` in `keys.ts` for plugging in Mastra-compatible language models, with detailed documentation for application integration. [[1]](diffhunk://#diff-94a3c27d4b6836bf7512e4760b142f9abc4375a2fe001cacfb72b171c165f102R4) [[2]](diffhunk://#diff-94a3c27d4b6836bf7512e4760b142f9abc4375a2fe001cacfb72b171c165f102R59-R80)

**ChatStore Enhancements for Mastra Compatibility**
- Added Mastra-compatible methods to `ChatStore` for adding human, AI, and tool messages using plain strings, and for retrieving messages and converting them to a Mastra-compatible format (`CoreMessageLike`). These methods avoid LangChain types and facilitate seamless workflow migration. [[1]](diffhunk://#diff-cf467422127bc256cdb1ec2794f35c2f0a7fa0830ff397c171849f3e9be4ee1cR18) [[2]](diffhunk://#diff-cf467422127bc256cdb1ec2794f35c2f0a7fa0830ff397c171849f3e9be4ee1cR116-R215) [[3]](diffhunk://#diff-cf467422127bc256cdb1ec2794f35c2f0a7fa0830ff397c171849f3e9be4ee1cR321-R380)

**Test Refactoring for WorkflowRunner**
- Refactored integration tests for `GenerationService` to use `WorkflowRunner` and async generators, improving test reliability and aligning with the new Mastra workflow execution model. Utility functions for empty and error event streams were added for robust testing. [[1]](diffhunk://#diff-1508ed91a3800267548c8dbd513552b1472bc2d4a2740946d13c27403261e5dcL1-R40) [[2]](diffhunk://#diff-1508ed91a3800267548c8dbd513552b1472bc2d4a2740946d13c27403261e5dcL33-R88) [[3]](diffhunk://#diff-1508ed91a3800267548c8dbd513552b1472bc2d4a2740946d13c27403261e5dcL153-R121)

**Event Type Extensions**
- Extended `LLMStreamEvent` in `event.types.ts` to include a new `LLMStreamErrorEvent` type, improving error handling and event stream expressiveness.

### CODE REVIEW

**Review by rohit-sourcefuse — CHANGES_REQUESTED** [review]
## Review: feat(migration): LangGraph → Mastra Phase 1

Good direction overall. The typed `WorkflowRequestContext` bridge, function-based model injection on the Agent, and `ContextWindowManager` are all well-designed. However there are several blockers that prevent this from functioning at runtime, plus two correctness bugs in the concurrency layer.

---

### BLOCKERS — must fix before merge

**1. `run.stream()` not awaited (`workflow-runner.ts:111`)**

`run.stream()` returns `Promise<{stream: AsyncIterable<...>}>`. The result is not awaited and `stream` is not destructured — `workflowStream` holds a Promise, not an AsyncIterable. `workflowStream[Symbol.asyncIterator]()` throws `TypeError` on every chat request. TypeScript misses this because `@mastra/core` is not installed (return type is `any`).

```typescript
// Fix:
const run = chatWorkflow.createRun();
const {stream: workflowStream} = await run.stream({
  inputData: {prompt, files, sessionId},
  requestContext,
});
yield* this._mergeStreams(workflowStream, eventQueue, abortController);
```

**2. `AsyncEventQueue` single-slot resolver race condition (`async-event-queue.ts:17`)**

`_resolve` is a single slot. If two pushes happen before the consumer resumes (two simultaneous tool calls), the second `push()` overwrites `_resolve = null` and drops the first wakeup. The `while(true)` loop in the iterator stalls permanently — the SSE stream hangs silently.

```typescript
// Fix: replace _resolve: (() => void) | null with an array
private readonly _resolvers: Array<() => void> = [];

push(event: LLMStreamEvent): void {
  if (this._closed) return;
  this._queue.push(event);
  this._resolvers.splice(0).forEach(r => r());
}

close(): void {
  this._closed = true;
  this._resolvers.splice(0).forEach(r => r());
}

// In iterator:
await new Promise<void>(resolve => { this._resolvers.push(resolve); });
```

**3. `@mastra/core` not installed**

`node_modules/@mastra` does not exist. Every import from `@mastra/core/workflows`, `@mastra/core/agent`, `@mastra/core/request-context` fails with `Cannot find module`. Both CI checks fail for this reason. Run `npm install`, confirm subpath exports resolve, commit the updated lock-file.

**4. `@langchain/openrouter` does not exist (`openrouter.provider.ts:2`)**

`@langchain/openrouter` is not a published npm package. This new sub-module does not compile. Either remove it from this PR or implement via `@ai-sdk/openai` with `baseURL: 'https://openrouter.ai/api/v1'`.

**5. `@langchain/classic/vectorstores/memory` does not exist (`inmemory.vector.ts:2`)**

No such package. Should be `@langchain/community/vectorstores/memory`.

**6. `LLMStreamErrorEvent.data.message` breaks the SSE contract (`event.types.ts:65`)**

The new type uses `{ data: { message: string } }`. The TDD spec (§11.4) and existing client expectations use `{ data: { error: string } }`. Any client checking `event.data.error` receives `undefined` silently. Change the field to `error: string` and update the emit site in `agent-reasoning.step.ts`.

---

### IMPORTANT — address before Phase 2

**7. `_mergeStreams` leaks unhandled promise rejections (`workflow-runner.ts:131`)**

When `Promise.race()` settles on one source, the other pending `.next()` promise is abandoned. If it later rejects, Node.js emits `UnhandledPromiseRejection`. Add a `try/finally` that calls `wsIter.return?.()` and `qIter.return?.()` to properly close both iterators on exit.

**8. Log events silently dropped (`workflow-runner.ts:180, 202`)**

`LLMStreamEventType.Log` events are filtered out from both sources. The existing LangGraph system forwards Log events — they drive "Reading file: ..." and "Context compressed: ..." UI messages. Dropping them silently is a backward-compat regression. Remove the filter or document it explicitly as a breaking change.

**9. Unused `@repository(ChatRepository)` injection (`workflow-runner.ts:73`)**

Injected but never used anywhere in the class body. Remove it.

**10. Dead LangGraph services still registered (`component.ts:87-103`)**

`CallLLMNode`, `RunToolNode`, `InitSessionNode`, `SummariseFileNode`, `ContextCompressionNode`, `EndSessionNode`, `ChatGraph` are all still in `this.services`. Since `GenerationService` now uses `WorkflowRunner` exclusively, these nodes are instantiated per request but never called. Remove them or gate with a rollback feature flag.

**11. Silent tool message loss (`persist-conversation.step.ts:52-70`)**

When `addAIMessageText()` returns `undefined`, the `if (aiMessage)` guard silently skips persisting all tool call records. Throw explicitly instead:
```typescript
if (!aiMessage) {
  throw new Error(
    `PersistConversationStep: failed to persist AI message for session ${sessionId}`,
  );
}
```

**12. `AiIntegrationBindings.Checkpointer` not marked deprecated (`keys.ts:37`)**

Per TDD §26.8 this key has no Mastra equivalent and should be `@deprecated` now.

---

### Test coverage

The only new test file stubs `WorkflowRunner` and verifies call counts — it provides no regression protection for the new Mastra layer. Zero tests for: `AsyncEventQueue` (push/close/concurrency), `ContextWindowManager` (system-message-preserved, array-typed content, boundary conditions), `TokenUsageAccumulator`, `WorkflowRunner._mergeStreams` (abort, one-source-exhausted), all 6 workflow steps, and `ChatStore.toCoreMessage()` (especially the `tool` role branch that constructs the `tool-result` content array passed directly to the LLM API).

`AsyncEventQueue` and `ContextWindowManager` are dependency-free pure classes — adding tests for them would have caught the resolver race before review.

---

**Must fix before merge:** items 1–6
**Fix before Phase 2:** items 7–12 + unit tests for pure utility classes

File: src/mastra/bridge/workflow-runner.ts [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+    requestContext.set('systemContext', this.systemContext);
+    requestContext.set('tokenUsageAccumulator', tokenAccumulator);
+
+    const run = await chatWorkflow.createRun();
+
+    // run.stream() executes the workflow lazily as we consume the returned iterator.
+    // The iterator yields WorkflowStreamEvent — steps emit via writer.write() which
+    // surfaces as {type: 'workflow-step-output', payload: {output: <our event>}}.
+    const workflowStream = run.stream({
+      inputData: {prompt, files, sessionId},
```
Comment (rohit-sourcefuse) [inline src/mastra/bridge/workflow-runner.ts]: Test comment - position 112

**Review by rohit-sourcefuse — COMMENTED** [review]
test

**Review by rohit-sourcefuse — COMMENTED** [review]
.

File: src/mastra/bridge/async-event-queue.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * Architecture:
+ *   1. WorkflowRunner stores an AsyncEventQueue in RequestContext before workflow run.
+ *   2. AgentReasoningStep's onStepFinish / tool-call callbacks push events to the queue.
+ *   3. WorkflowRunner reads from the queue concurrently and sends events to ITransport.
+ *   4. After workflow completes, `close()` signals the consumer to stop iterating.
+ */
+export class AsyncEventQueue implements AsyncIterable<LLMStreamEvent> {
+  private readonly _queue: LLMStreamEvent[] = [];
+  private _resolve: (() => void) | null = null;
+  private _closed = false;
```
Comment (rohit-sourcefuse) [inline src/mastra/bridge/async-event-queue.ts:18]: Test comment - position 18

File: src/graphs/event.types.ts [hunk 3]
```diff
@@ -61,11 +61,19 @@ export type LLMStreamInitEvent = {
   };
 };
 
+export type LLMStreamErrorEvent = {
+  type: LLMStreamEventType.Error;
+  data: {
```
Comment (rohit-sourcefuse) [inline src/graphs/event.types.ts:66]: Test comment - position 6

**Review by rohit-sourcefuse — COMMENTED** [review]
test

File: src/mastra/bridge/workflow-runner.ts (hunk 1 above)
Comment (rohit-sourcefuse) [inline src/mastra/bridge/workflow-runner.ts]: **[BLOCKER] `run.stream()` is not awaited — runtime crash on every request**

Per `@mastra/core ^1.32.1`, `run.stream()` returns `Promise<{stream: AsyncIterable<WorkflowStreamEvent>}>`. The result must be awaited and the `stream` property destructured. As written `workflowStream` holds a Promise — `workflowStream[Symbol.asyncIterator]()` throws `TypeError` on every request.

```typescript
const run = chatWorkflow.createRun();
const {stream: workflowStream} = await run.stream({
  inputData: {prompt, files, sessionId},
  requestContext,
});
yield* this._mergeStreams(workflowStream, eventQueue, abortController);
```

File: src/mastra/bridge/workflow-runner.ts [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    private readonly toolStore: ToolStore,
+    @inject(AiIntegrationBindings.Config, {optional: true})
+    private readonly aiConfig: AIIntegrationConfig | undefined,
+    @inject(AiIntegrationBindings.SystemContext, {optional: true})
+    private readonly systemContext: string[] | undefined,
+    @inject.getter(AuthenticationBindings.CURRENT_USER)
+    private readonly getCurrentUser: Getter<IAuthUserWithPermissions>,
+    @repository(ChatRepository)
+    private readonly chatRepository: ChatRepository,
+  ) {}
```
Comment (rohit-sourcefuse) [inline src/mastra/bridge/workflow-runner.ts:139]: **Unused injection — remove `chatRepository`**

`@repository(ChatRepository)` is injected here but never referenced in the class body. Remove it to eliminate the dead DI dependency.

**Review by rohit-sourcefuse — COMMENTED** [review]
Inline comments on specific issues. See the full request-changes review above for the complete list.

File: src/mastra/bridge/workflow-runner.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+          // Extract LLMStreamEvent from workflow-step-output chunks
+          const chunk = result.value as {
+            type?: string;
+            payload?: {output?: unknown};
+          };
+          if (chunk?.type === 'workflow-step-output') {
+            const output = chunk.payload?.output;
+            if (isLLMStreamEvent(output)) {
+              if (output.type !== LLMStreamEventType.Log) {
+                yield output;
```
Comment (rohit-sourcefuse) [inline src/mastra/bridge/workflow-runner.ts:613]: **Log events silently dropped — backward-compat regression**

The LangGraph system forwarded `Log` events to clients, driving 'Reading file: ...' and 'Context compressed: ...' UI messages. Filtering them here without a changelog entry silently breaks that behavior. Remove the filter or document it explicitly as a breaking change.

File: src/mastra/bridge/async-event-queue.ts (hunk 2 above)
Comment (rohit-sourcefuse) [inline src/mastra/bridge/async-event-queue.ts:18]: **[BLOCKER] Single-slot resolver — SSE stream stalls on concurrent tool calls**

`_resolve` is a single slot. If two `push()` calls happen before the consumer resumes (parallel tool calls), the second call drops the first wakeup and the `while(true)` loop stalls permanently — SSE stream hangs silently with no error.

Replace with an array:

```typescript
private readonly _resolvers: Array<() => void> = [];

push(event: LLMStreamEvent): void {
  if (this._closed) return;
  this._queue.push(event);
  this._resolvers.splice(0).forEach(r => r());
}

close(): void {
  this._closed = true;
  this._resolvers.splice(0).forEach(r => r());
}

// In iterator:
await new Promise<void>(resolve => { this._resolvers.push(resolve); });
```

File: src/graphs/event.types.ts (hunk 3 above)
Comment (rohit-sourcefuse) [inline src/graphs/event.types.ts:66]: **[BLOCKER] Error event field name breaks SSE contract**

The field is `message: string` here, but the TDD spec (§11.4) and existing client expectations use `error: string`. Any client checking `event.data.error` receives `undefined` silently.

```typescript
export type LLMStreamErrorEvent = {
  type: LLMStreamEventType.Error;
  data: { error: string };  // not 'message'
};
```

Update the emit site in `agent-reasoning.step.ts` to match: `data: {error: errMsg}`.

File: src/mastra/workflows/chat/steps/persist-conversation.step.ts [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+    debug(
+      `PersistConversation: session=${sessionId}, textLen=${finalText.length}, tools=${toolCalls.length}`,
+    );
+
+    // 1. Persist the AI's text response
+    const aiMessage = await chatStore.addAIMessageText(sessionId, finalText);
+
+    if (!aiMessage) {
+      debug('PersistConversation: addAIMessageText returned undefined');
```
Comment (rohit-sourcefuse) [inline src/mastra/workflows/chat/steps/persist-conversation.step.ts]: **Silent tool message loss when `addAIMessageText` returns `undefined`**

If the AI message persistence fails, the `if (aiMessage)` guard on line 70 silently skips all tool call records for the turn. Data is lost with only a debug log.

```typescript
if (!aiMessage) {
  throw new Error(
    `PersistConversationStep: failed to persist AI message for session ${sessionId}`,
  );
}
```

File: src/component.ts [hunk 7]
```diff
@@ -88,6 +89,8 @@ export class AiIntegrationsComponent implements Component {
       TokenCounter,
       GenerationService,
       ChatStore,
+      // mastra migration
+      WorkflowRunner,
       // graph
```
Comment (rohit-sourcefuse) [inline src/component.ts]: **Dead LangGraph services still registered — remove or gate with a rollback flag**

Since `GenerationService` now delegates entirely to `WorkflowRunner`, all six LangGraph nodes and `ChatGraph` in this `services` array are instantiated per request by LB4 DI but never called. Remove them, or keep them behind a `WORKFLOW_ENGINE=langgraph` feature flag for rollback safety (TDD §22.5).

File: src/keys.ts [hunk 8]
```diff
@@ -55,6 +56,28 @@ export namespace AiIntegrationBindings {
   export const SystemContext = BindingKey.create<string[]>(
     `services.ai-reporting.system-context`,
   );
+
+  // ── Mastra LLM bindings (Phase 1 migration) ──────────────────────────────
+  /**
```
Comment (rohit-sourcefuse) [inline src/keys.ts:100]: **Mark `Checkpointer` as `@deprecated` now (TDD §26.8)**

This binding has no Mastra equivalent and will be removed in Phase 5. Consumers need advance notice before `@langchain/langgraph` is dropped.

```typescript
/**
 * @deprecated No Mastra equivalent. Never wired in graph.compile({}).
 * Will be removed in Phase 5 when @langchain/langgraph is dropped.
 */
export const Checkpointer = ...
```
